### ISSUE TITLE
Silent crashes

### ISSUE THREAD

**Issue by vedat-keklik** [galkahana/HummusJS#118]
Hi,

I noticed that, library becomes unresponsive under certain circumstances. For example, if I give wrong file path for a font (`__dirname + '/fonts/nonexistent.ttf'`), or that file is inaccessible for a reason then execution stops, but process keeps running silently with high resource consumption. I didn't intend to break something. My own tests warned me about that. I saw this behavior because of another input errors too (also some invalid pdf commands). 

I am doing **"file exists?"** check. Hence, I can avoid **file access** errors. But the problem is library doesn't say me anything about "what is wrong"  **sometimes** and the process becomes a zombie one.

Do you have any idea?

Thanks.

EDIT:

My environment:

Ubuntu 16.04 64-bit
Node v6.6.0

**Comment by galkahana** [comment]
Hi @vedat-keklik . while i did make sure to provide error codes and the ability to know when something is wrong with the C++ library, when the time came for the JS driver i invested less in accounting for returned error codes and more with providing as much of the API as possible. As a result i would expect that invalid inputs (like providing an non existent file) may cause errors. as for why they create zombie processes, this has to do with how the OS/NodeJS implementation deal with such unexpected. I guess one could provide a layer on top of hummus making it safer. at the time i preferred performance over validations.

**Comment by vedat-keklik** [comment]
Thanks for the answer. 

It is ok. I am validating inputs as much as i can. I don't think that input validation is your job. But it would be great if you can add a "global" exception handler. This may sound silly, because I am unfamiliar with binding C++ code with node.js. Just a thought. 

By the way, thanks for Hummus project. There is no JS alternative.

**Comment by galkahana** [comment]
@vedat-keklik thanks for your note.
I'm curious - what makes you say "there's no JS alternative". I know that most people use phantom as method to generate PDF files. also, what about mozilla's (pdf.js)[http://mozilla.github.io/pdf.js/]?

**Comment by vedat-keklik** [comment]
I said that because, Hummus can **modify** PDF files. Alternatives that you've mentioned are focused on **generating PDF files from scratch** or converting them to another document formats as far as I know.  In fact, there is a project [npm/scissors](https://www.npmjs.com/package/scissors) for modification. But it requires 3rd party software install (pdftk). Also I needed more than scissors' API. 

To sum up, I am using Hummus for its **modification** capabilities. ***There is no JS alternative that can do this.*** Or there is, but I don't know them.

I am not sure that other people really care about 'modifying PDF files'. But it is really hard to do if you don't have tools. You need to parse file, do many other things...

Thanks again.

**Comment by galkahana** [comment]
Makes sense. thank you.
